### PR TITLE
pybind: add param option to argdesc in ceph_argparse.py

### PR DESCRIPTION
**Pull request by matthewoliver**
This patch adds an new parameter to the argdesc class, `param` that when
set to `true` will force the `--name=` syntax in the generated helpstr.

The `param=true` is also added to the `fomat` option of the
`ceph orchestrator device ls` command so the specified usage wont result
in a cli syntax error.

This patch also modifies the ceph client to remove any param=true
argdesc's from the srgs and deals with them non positionally.

Fixes: https://tracker.ceph.com/issues/42113
Signed-off-by: Matthew Oliver <moliver@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by matthewoliver** [comment]
Wasn't 100% sure of what module it I should prefix it with. the bug is from the RADOS project in redmine, so used that.

Also I noticed there wasn't any tests for the helpstr at all. So no tests in this patch, I could go create some if that helps. Though that could be a long piece of string :P

**Comment by smithfarm** [comment]
@matthewoliver Maybe "pybind" would be a better component prefix?

**Comment by matthewoliver** [comment]
Cool, thanks, will update it in the morning.

**Comment by sebastian-philipp** [comment]
Hm. `format` is special, as it is typically not explicitly added to the command specification. Instead it is assumed that all commands support the `--format` parameter, which is not the case. Maybe we can just remove the line

https://github.com/ceph/ceph/pull/31042/files#diff-b3018215c5634fd57e880aaec0a79dcaL81

altogether?

**Comment by matthewoliver** [comment]
> @matthewoliver Maybe "pybind" would be a better component prefix?

done

**Comment by matthewoliver** [comment]
@sebastian-philipp Maybe, but the problem I see is that the python side only deals with json or plain.
The `--help` command does mention the `-f/--format` option, so one could argue that this it isn't really required in the usage side?:

  `orchestrator device ls [<host>...] [--format=json|plain] [--refresh]                                                   List devices on a node`

Although, the `--help` text (non usage section) does mention the format options being "plain, json, json-pretty, xml and xml-pretty" whereas the orchestrator cli, well the python side, certainly doesn't support all those, only json and plain, which is probably why it was added to the usage side in the first place.

So the advantage of having it where it is means it shows that the `ceph orchestrator device ls --format=..` command only supports a subset of the whole, and which subset. If we remove it then the command not supporting xml etc could be considered a bug.

**Comment by sebastian-philipp** [comment]
Thing is, converting json to yaml or xml or xml-pretty or json-pretty is rather trivial. Heck, we could even convert json to plain [automatically](https://gist.github.com/sebastian-philipp/8265059ea9241e2e8b65de0652e6a81f). Having **all** commands to reply different formats independently is not the way I want this to work. I'd really like to get rid of all the manual format handling in mgr commands instead of improving (and fixing) my hack to manually support `--format`.

**Comment by liewegas** [comment]
Setting aside the --format part in particular (which is perhaps a special-case), I like the idea of param=True to specify arguments that must be use ``--foo[= ]bar``.  My request would that ``ceph.in`` is also update so that these are *not* parsed positionally (as they are now).

**Comment by tchaikov** [comment]
@matthewoliver ping?

**Comment by matthewoliver** [comment]
Sorry, will loop back to this. Kinda moved on to some other patches.. Sorry. I think I understand what Sage is saying. I'll have a poke around to be sure :)

**Comment by matthewoliver** [comment]
Just a rebase. Looking in to Sage's comment now.

**Comment by liewegas** [comment]
This is great.  I think there is a missing piece, though: the ``src/ceph.in`` change to disallow a param as a positional argument.  (I'm a little worried that will be a slightly hairy change since I think the positional logic has some cleverness to infer what a positional argument maps to based on its type?  But hopefully it's a straightforwad change.)

File: src/pybind/ceph_argparse.py [hunk 1]
```diff
@@ -1023,6 +1082,7 @@ def validate(args, signature, flags=0, partial=False):
                 if kwarg_desc:
                     validate_one(kwarg_v, kwarg_desc)
                     store_arg(kwarg_desc, d)
+                    matchcnt += 1
```
Comment (matthewoliver) [inline src/pybind/ceph_argparse.py]: This is really just a driveby fix, if you pass a long (--) paramenter using it's name in the CLI without param=true. Then it'll be checked positionally. The cmdict ends up correct and a with the right matches but because we don't increment the matchcnt, the sig being validated will fail.

**Comment by matthewoliver** [comment]
This version, deals with any param=true argdesc's non positionally before we check the positional args.
I've only added this to the param=true at the moment, because I don't want to mess too much with the arg parsing. If you like how I pulled them out, we could refactor 'validate' some more and always pull out  any long params to they aren't positionally checked. But maybe we could do that in a follow up patch as to not blow out the scope of this bug fix anymore :)

As it stands if you want to pull out any long params just add a 'param=true' to them.

Eitherway, let me know.. happy to scope creeep ;P

**Comment by matthewoliver** [comment]
Been thinking a little more about this. I wonder if it would be worth writing some code that converts the signatures returned by get_command_descriptions into argparse parsers, or maybe even have a CephArgumentPaser that can take in a sig and set it up? 
Thinking about this, python's argparse I'm pretty sure can handle args and commands (ie sub commands). From what I see argdesc's that are prefix could just become subcommands.

I guess the way we deal with args positionally may make this impossible though, ie sending in something with or without the --<name> options.
We could roll them together, ie look for --<name> args and pull them out non positionally, adjust the sigs, and run whatever left positionally. argparse makes the former simpler (as you can see from this patch for the param=true). 

Anyway, just brain storming.

**Comment by tchaikov** [comment]
@mattbenjamin just my two cents,

1. i think param should be be renamed to positional, which should be an implicit property of a param in the command specification. and it defaults to "true".
2. i think we should reset "positional" to "false" if any parameter is located after a parameter which accepts a variable number of arguments. we could do this in `parse_funcsig()`.

**Comment by sebastian-philipp** [comment]
In the context of #32826 it would be actually really helpful!

**Comment by liewegas** [comment]
:+1: agree!

**Comment by liewegas** [comment]
How about this:

- we go with this change for now (either with or without the minor changes @tchaikov suggests: renaming param to posiitonal and implicitly setting it after n=N args)
- post-octopus, we consider what changes would be needed to turn these sigs into argparse parsers.  (IIUC the  main stumbling block is that you can currently specify two position arguments with req=false of different types and we'll disambiguate which you meant; that probably needs to go away for argparse, and probably should go away for the sake of removing ambiguity from our CLI.)

**Comment by sebastian-philipp** [comment]
needs rebase
